### PR TITLE
Migrate masked_scatter to use cub instead of thrust

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Benchmark:

```python
import torch
import itertools


def run50_sync(f):
    for _ in range(50):
        f()
    torch.cuda.synchronize()

    
run50_sync(lambda: torch.randperm(1000000, device='cuda'))

    
def benchmark(M):
    a = torch.randn(M, device='cuda')
    m = torch.randint(1, (M,), dtype=torch.long, device='cuda').bool()
    v = torch.randn(M, device='cuda')

    torch.cuda.synchronize()

    %timeit run50_sync(lambda:a.masked_scatter_(m, v))

for M in (100, 1000, 100000, 10000000):
    print(M)
    benchmark(M)
```

Before:
```
100
8.65 ms ± 80.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
1000
8.75 ms ± 72.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
100000
9.27 ms ± 87.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
10000000
33.6 ms ± 358 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

After
```
100
8.04 ms ± 37.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
1000
8.09 ms ± 38.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
100000
8.63 ms ± 76.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
10000000
31.9 ms ± 298 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

### CODE REVIEW

File: aten/src/ATen/native/cuda/IndexKernel.cu [hunk 1]
```diff
@@ -385,20 +382,9 @@ void masked_scatter_cuda_impl(Tensor& self, const Tensor& mask, const Tensor& so
...
-
-  // Reference for using static_cast on `init_value`:
-  // https://github.com/NVIDIA/thrust/issues/1379
-  thrust::exclusive_scan(
-      thrust::cuda::par(allocator).on(c10::cuda::getCurrentCUDAStream()),
-      maskData,
-      maskData + mask_cont.numel(),
-      maskPrefixSumData,
-      static_cast<int64_t>(0));
+  at::cuda::cub::exclusive_sum(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/IndexKernel.cu:385]: mask_cont.numel() > int_max used to be supported w/o extravagant memory requirements, so probably we shouldn't remove this support? It's slightly annoying to add it (see cumsum where it's done for inclusive sum), but it's doable.

File: aten/src/ATen/native/cuda/IndexKernel.cu [hunk 2]
```diff
@@ -385,20 +381,9 @@ void masked_scatter_cuda_impl(Tensor& self, const Tensor& mask, const Tensor& so
   // Use a prefix sum to determine the output locations of the masked elements
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/IndexKernel.cu:381]: Can you please also move `totalElements` calculation here (line 373), and use direct cub calls, since `mask` is made contiguous anyway. Turns out mask.sum() first converts mask to long, and then reduces long tensor. Using cub directly should be much better.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/IndexKernel.cu:374]: done

File: aten/src/ATen/cuda/cub.cuh [hunk 3]
```diff
@@ -151,4 +153,125 @@ static inline void segmented_sort_pairs(
...
+  int size_cub = std::min<int64_t>(num_items, max_cub_size);
+  CUB_WRAPPER(NO_ROCM(detail)::cub::DeviceScan::ExclusiveScan,
+      input,
+      output,
+      scan_op,
+      init_value,
+      size_cub,
+      at::cuda::getCurrentCUDAStream());
+  C10_CUDA_KERNEL_LAUNCH_CHECK();
+  auto allocator = c10::cuda::CUDACachingAllocator::get();
```
Comment (ngimel) [inline aten/src/ATen/cuda/cub.cuh:255]: Can you move this inside the loop, so that small inputs don't have to pay the allocation price? Large inputs could potentially do it repeatedly, but it will be hidden behind massive kernels.

Comment (zasdfgbnm) [reply aten/src/ATen/cuda/cub.cuh]: fixed

File: aten/src/ATen/native/cuda/IndexKernel.cu [hunk 4]
```diff
@@ -373,32 +369,22 @@ template <typename mask_t>
...
+  at::cuda::cub::exclusive_scan(
+    mask_cont.data_ptr<mask_t>(), maskPrefixSum.data_ptr<int64_t>(),
+    []__device__(int64_t a, int64_t b) { return a + b; }, int64_t(0),
+    mask_cont.numel());
 
-  thrust::device_ptr<mask_t> maskData(mask_cont.data_ptr<mask_t>());
-  thrust::device_ptr<int64_t> maskPrefixSumData(
-      maskPrefixSum.data_ptr<int64_t>());
+  // Determine our output size
+  auto totalElements = (maskPrefixSum[-1] + mask_cont[-1]).item<int64_t>();
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/IndexKernel.cu:383]: what's mask_cont[-1] for a multi-d mask?

**Review by ngimel — APPROVED** [review]
